cpp/include/games/stochastic_nim/PerfectPlayer.hpp

```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+#include <core/AbstractPlayer.hpp>
+#include <core/BasicTypes.hpp>
+#include <games/stochastic_nim/Constants.hpp>
+#include <games/stochastic_nim/Game.hpp>
+#include <util/BitSet.hpp>
+#include <util/EigenUtil.hpp>
+
+namespace stochastic_nim {
+
+class PerfectStrategy {
+ public:
+  struct Params {
+    const int starting_stones;
+    const int max_stones_to_take;
+    const float* chance_event_probs;
+    const int num_chance_events;
+  };
+
+  PerfectStrategy(Params params);
+  float* get_state_value() const { return state_value_; }
+  int* get_optimal_action() const { return optimal_action_; }
+  Params get_params() const { return params_; }
+
+ private:
+  void iterate();
+  Params params_;
+  // expected win rate if there are [index up to starting_stones] stones left after a player's move
+  float* state_value_;
+  // best number of stones to take if there are [index up to starting stones] stones left
+  int* optimal_action_;
```

Comment:
If directly using the values in `Constants.hpp`, you can simply declare these as,

```
float state_values_[N];
int optional_actions_[N];
```

Note that you should pluralize for better grammar.